buffer/src/main/java/io/netty5/buffer/internal/CleanerPool.java

```diff
@@ -0,0 +1,200 @@
+     * Cleaners are distributed in round robin, and the same EventLoop will get the same Cleaner for every
+     * nextCleaner method invocations.
+     */
+    private final CleanerThreadLocal threadLocalCleaner = new CleanerThreadLocal(cleaners);
+
+    /**
+     * Static initializer used to setup Method Handles that are used to obtain loom virtual Threadfactory
+     * (if the platform supports it).
+     */
+    static {
+        MethodHandle mhOfVirtualTmp = null;
+        MethodHandle mhFactoryTmp = null;
+
+        if (USE_VTHREAD) {
+            try {
+                Class<?> clzOfVirtual = Class.forName("java.lang.Thread$Builder$OfVirtual");
+                mhOfVirtualTmp = publicLookup.findStatic(Thread.class, "ofVirtual",
+                        MethodType.methodType(clzOfVirtual));
+                mhFactoryTmp = publicLookup.findVirtual(clzOfVirtual, "factory",
+                        MethodType.methodType(ThreadFactory.class));
+            } catch (ClassNotFoundException e) {
+                logger.debug("Loom not supported, Cleaner will use default cleaner daemon threads.");
+            } catch (Throwable t) {
+                logger.warn("Could not create virtual thread factory, will use default cleaner daemon threads.", t);
+            }
+        }
+
+        mhOfVirtual = mhOfVirtualTmp;
+        mhFactory = mhFactoryTmp;
+    }
+
+    /**
+     * The singleton for the CleanerPool.
+     */
+    public static final CleanerPool INSTANCE = new CleanerPool();
+
+    /**
+     * The FastThreadLocal used to map a Cleaner instance to Event Loop threads.
+     */
+    private static class CleanerThreadLocal extends FastThreadLocal<Cleaner> {
+        private final Cleaner[] cleaners;
+        private static final AtomicInteger counter = new AtomicInteger();
+
+        CleanerThreadLocal(Cleaner[] cleaners) {
+            this.cleaners = cleaners;
+        }
+
+        @Override
+        protected Cleaner initialValue() {
+            return cleaners[counter.getAndIncrement() % POOL_SIZE];
+        }
+    }
+
+    private CleanerPool() {
+        logger.info("Instantiating CleanerPool: {}={}, {}={}, using vthreads={}",
+                        CNF_POOL_SIZE, POOL_SIZE,
+                        CNF_USE_VTHREAD, USE_VTHREAD,
+                        mhFactory != null);
+    }
+
+    /**
+     * Returns the next available cleaner to the calling thread.
+     * The same thread will get the same cleaner for every nextCleaner method calls.
+     */
+    Cleaner nextCleaner() {
+        return threadLocalCleaner.get();
+    }
+
+    private static int getPoolSize() {
+        int poolSize = SystemPropertyUtil.getInt(CNF_POOL_SIZE, 0);
+        if (poolSize < 0) {
+            throw new IllegalArgumentException(CNF_POOL_SIZE + " is negative: " + poolSize);
+        }
+        return poolSize == 0 ? Runtime.getRuntime().availableProcessors() : poolSize;
```

Comment:
Thanks for the feedback, Since I have applied the suggestion made by @franz1981, we have now all event-loops having their own Cleaner instance automatically. So no need to derive the pool size from the actual number of event loop threads I think.  
And for external threads, as said Francesco, if users would use external threads, make sense they’ll pay some contention, and in this case they can adjust the -Dio.netty5.cleanerpool.size system property (which is 1 by default). The pool is allocated lazily (it won't if no external threads).

Reply:
loving it, well done!